src/server/service/batch.rs

```diff
@@ -83,60 +93,124 @@ impl ReqBatcher {
     }
 }
 
+pub struct BatcherBuilder {
+    pool_size: usize,
+    enable_batch: bool,
+}
+
+impl BatcherBuilder {
+    pub fn new(enable_batch: bool, pool_size: usize) -> Self {
+        BatcherBuilder {
+            enable_batch,
+            pool_size,
+        }
+    }
+    pub fn build(&self, queue_per_worker: usize, req_batch_size: usize) -> Option<ReqBatcher> {
+        if !self.enable_batch {
+            return None;
+        }
+        if req_batch_size > self.pool_size * MIN_BATCH_GET_REQUEST_COUNT
+            && queue_per_worker >= MIN_BATCH_GET_REQUEST_COUNT
+        {
+            return Some(ReqBatcher::new(req_batch_size / self.pool_size));
+        }
+        if req_batch_size >= MIN_BATCH_GET_REQUEST_COUNT
+            && queue_per_worker >= MAX_QUEUE_SIZE_PER_WORKER
```

Comment:
I'm OK with this algorithm, but did you run benchmark using `req_batch_size / self.pool_size` without considering `MIN_BATCH_REQUEST_COUNT` or `queue_per_worker`? I just can't think of a losing case, assuming requests are evenly dispatched to multiple queues. E.g. 8 requests with 8 queues = no batch, 8 requests with 4 queues = batch 2.